Make Popup more modal.

grokys opened the pull request:
## What does the pull request do?

As described in #3965, mouse move etc events are still passed to the parent window when a popup is open. This doesn't happen in other frameworks, e.g. WPF.

WPF uses a capture on the mouse but that isn't feasible in Avalonia as we need a pointer event in order to capture the pointer. First attempt used event filtering but that caused problems with tooltips getting "stuck" (see comments) so this PR now uses an overlay layer:

- [x] Added `Popup.IsLightDismissEnabled` property (from UWP) and deprecate `StaysOpen` in favor of this property
- [x] Added `Popup.OverlayDismissEventPassThrough` to control whether dismiss events are passed through from the overlay layer to the underlying window content
  - The previous method of controlling this for `ComboBox` (introduced in #3857) will no longer work. Instead the `OverlayDismissEventPassThrough` property should be set on the `Popup` in the `ComboBox` template
- [x] Added `LightDismissOverlayLayer.InputPassThroughElement` which implements something similar to UWP's `FlyoutBase.OverlayInputPassThroughElement`. With this feature, `Menu`/`MenuItem` can use `IsLightDismissEnabled="True"`
  - This is needed because the top-level menus on the main window still need to receive input events even when the menu popup is open
  - Implemented by adding a `ICustomHitTest` which is similar to the existing `ICustomSimpleHitTest` but is recognized by the deferred renderer as well as the immediate renderer
  - ~~We might want to also add this to `Popup` but given that in UWP it's defined on `FlyoutBase` instead of `Popup` and that `Menu` goes direct to the overlay layer, I haven't done this in this PR~~ Added as `Popup.OverlayInputPassThroughElement`.

Also added nullable annotations to the files I touched.

## Fixed issues

Fixes #3965

grokys commented:
@kekekeks mentioned in chat this he doesn't like the filtering raw input method of doing this, and I'm inclined to agree. If we don't use filtering, what are our options? The options that come to mind are:

1. Require that popups be opened as a result of interaction and pass an `IPointer`, we can then capture on this pointer.
  - Can't open popups without interaction (e.g. on a timer)
  - We can only capture events from the device that initiated the event
2. Add an overlay layer to the parent window/popup which to disable the window/popup contents
  - Looking at UWP, this may be the way it's done there
  - We could also implement a [`LightDismissOverlayMode`](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.primitives.popup.lightdismissoverlaymode?view=winrt-19041) like in UWP
  - Won't work if the parent window doesn't have a `VisualLayerManager`
  - Feels a bit hacky?

kekekeks commented:
>Require that popups be opened as a result of interaction and pass an IPointer, we can then capture on this pointer. 

Note that this won't work with touch pointers since pointer completely disappears when touch contact is lost.

maxkatz6 commented:
@grokys
Not sure if I cleary understand first option. Does it means that it will depend only on pointer? "Can't open popups without interaction (e. q. on timer)" sounds like "can't open popup programmatically". 
Popup should be able to be opened from keyboard (menu button) or something like TeachingTip control, which is Popup based, could be opened on any place of screen without pointer interaction. 
Also ToolTip is always delayed with timer. 

About second options. LightDismissOverlayMode hasn't special logic to confugure dismiss mode, it is just indicator, should it be darkened. You probably need [Popup.IsLightDismissEnabled](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.primitives.popup.islightdismissenabled?view=winrt-19041#Windows_UI_Xaml_Controls_Primitives_Popup_IsLightDismissEnabled)

grokys commented:
@maxkatz6 yeah I think option 1 is a no-go, so it's either the solution implemented here (raw event filtering) or option 2.

@kekekeks which do you prefer?

grokys commented:
> About second option. LightDismissOverlayMode is just indicator, should background be darkened. You probably need Popup.IsLightDismissEnabled

Yeah, I realize that. I was using `LightDismissOverlayMode` as a clue that UWP does in fact add a layer to disable the window contents.

MarchingCube commented:
@grokys @kekekeks Any progress on this matter? It is fairly important for user experience since otherwise we end up with random tooltips opening and pointerover effects that confuse people.

grokys commented:
No heard any more feedback on this, so going to merge it as-is if it works for @MarchingCube ?

grokys commented on src/Avalonia.Controls/Primitives/Popup.cs:
```diff
@@ -494,13 +494,25 @@ private void CloseCore(EventArgs? closeEvent)
-                CloseCore(e);
+                if (pointer.Root == _openState.TopLevel)
+                {
+                    // When StaysOpen == true, prevent move, wheel and leave window events being
```
Should be `StaysOpen == false`.

grokys commented:
As @MarchingCube told me in chat: this works fine for most popups but not menus as they use `StaysOpen= "True"`.

grokys commented:
Another problem with this:

- Button has a context menu and a tooltip
- Button is hovered, tooltip timer starts
- Context menu is opened
- Tooltip timer fires, tooltip is shown
- Tooltip doesn't get hidden even if mouse is no longer over the button because pointer events are being filtered out

I'm starting to think that a transparent overlay layer would be a better way to solve this...

grokys committed:
Show overlay layer when popup StaysOpen = false.

We need to stop events getting to parent window. Pointer capture won't work because this now needs an input event in order to take pointer capture, and filtering events on the parent window causes tooltips to continue tooltips to get stuck. Best solution would seem to be using an overlay layer.

grokys committed:
Fix failing popup tests.

grokys committed:
Added Popup.IsLightDismissEnabled property.

And deprecate `StaysOpen` in favor of this property.

grokys committed:
Added Popup.OverlayDismissEventPassThrough.

To control whether dismiss events are passed through from the overlay layer to the underlying window content.

grokys committed:
Make leak tests pass again.

grokys committed:
Nullable enable menu-related classes.

grokys committed:
Ensure that Open is called on main menu.

grokys committed:
Added LightDismissOverlayLayer.InputPassThroughElement.

Which implements something similar to UWP's `FlyoutBase.OverlayInputPassThroughElement`. With this feature, `Menu`/`MenuItem` can use `IsLightDismissEnabled="True"` fixing #3965 for top-level menus.

MarchingCube committed:
Merge branch 'master' into fixes/3965-popup-modal-interaction

MarchingCube reviewed (dismissed):
Functionality wise - LGTM. Covers my use cases but maybe other people can check as well.

grokys committed:
Added `Popup.OverlayInputPassThroughElement`.

MarchingCube commented:
Actually I will do more testing since there might be issues with pointer capture.

MarchingCube commented:
Looks like that clicking on overlay with input passthrough causes controls to get stuck in pressed state.

![overlay-passthrough-capture](https://user-images.githubusercontent.com/2588062/87721466-c0275580-c7b6-11ea-9ef6-f1c4aee5a968.gif)

grokys committed:
Merge branch 'master' into fixes/3965-popup-modal-interaction

grokys committed:
Added API compat changes.

grokys committed:
Set OverlayInputPassThroughElement in MenuItem template.

grokys committed:
Remove outdated submodule.

grokys committed:
Capture mouse when passing through event.

This moves the implicit capture from the overlay layer to the target element.

MarchingCube committed:
Merge branch 'master' into fixes/3965-popup-modal-interaction

MarchingCube reviewed (approved):
LGTM. Working great now.

grokys committed:
Merge branch 'master' into fixes/3965-popup-modal-interaction
